cpp/src/arrow/util/compression-test.cc

```diff
@@ -69,25 +71,303 @@ void CheckCodecRoundtrip(const vector<uint8_t>& data) {
+    bool should_retry = false;
+    do {
+      int64_t output_len = compressed.size() - compressed_size;
+      uint8_t* output = compressed.data() + compressed_size;
+      int64_t bytes_written;
+      ASSERT_OK(compressor->End(output_len, output, &bytes_written, &should_retry));
+      ASSERT_LE(bytes_written, output_len);
+      compressed_size += bytes_written;
+      if (should_retry) {
+        compressed.resize(compressed.capacity() * 2);
+      }
+    } while (should_retry);
+
+    compressed.resize(compressed_size);
+  }
+
+  // Then decompress...
+
+  std::vector<uint8_t> decompressed(2);
+  int64_t decompressed_size = 0;
+  {
+    const uint8_t* input = compressed.data();
+    int64_t remaining = compressed.size();
+
+    while (!decompressor->IsFinished()) {
+      // Feed a varying amount each time
+      int64_t input_len = std::min(remaining, make_buf_size());
+      int64_t output_len = decompressed.size() - decompressed_size;
+      uint8_t* output = decompressed.data() + decompressed_size;
+      int64_t bytes_read, bytes_written;
+      bool need_more_output;
+      ASSERT_OK(decompressor->Decompress(input_len, input, output_len, output,
+                                         &bytes_read, &bytes_written, &need_more_output));
+      ASSERT_LE(bytes_read, input_len);
+      ASSERT_LE(bytes_written, output_len);
+      ASSERT_TRUE(need_more_output || bytes_written > 0 || bytes_read > 0)
+          << "Decompression not progressing anymore";
+      if (need_more_output) {
+        decompressed.resize(decompressed.capacity() * 2);
+      }
+      decompressed_size += bytes_written;
+      input += bytes_read;
+      remaining -= bytes_read;
+    }
+    ASSERT_EQ(remaining, 0);
+    decompressed.resize(decompressed_size);
+  }
+
+  ASSERT_EQ(data.size(), decompressed.size());
+  ASSERT_EQ(data, decompressed);
+}
+
+class CodecTest : public ::testing::TestWithParam<Compression::type> {
+ protected:
+  Compression::type GetCompression() { return GetParam(); }
+
+  std::unique_ptr<Codec> MakeCodec() {
+    std::unique_ptr<Codec> codec;
+    ABORT_NOT_OK(Codec::Create(GetCompression(), &codec));
+    return codec;
+  }
+};
+
+TEST_P(CodecTest, CodecRoundtrip) {
   int sizes[] = {0, 10000, 100000};
   for (int data_size : sizes) {
     vector<uint8_t> data(data_size);
     random_bytes(data_size, 1234, data.data());
-    CheckCodecRoundtrip<CODEC>(data);
+    CheckCodecRoundtrip(GetCompression(), data);
+  }
+}
+
+TEST_P(CodecTest, StreamingCompressor) {
+  if (GetCompression() == Compression::SNAPPY) {
+    // SKIP: snappy doesn't support streaming compression
+    return;
+  }
+  if (GetCompression() == Compression::LZ4) {
+    // SKIP: LZ4 streaming compression uses the LZ4 framing format,
+    // which must be tested against a streaming decompressor
+    return;
+  }
+
+  int sizes[] = {0, 10, 100000};
+  for (int data_size : sizes) {
+    vector<uint8_t> data(data_size);
+    random_bytes(data_size, 1234, data.data());
```

Comment:
I would recommend testing on actual files (or very low entropy random) since uniform random data is not very compressible, thus there's less chance you trigger the buffer errors (if any) since input_size ~ output_size.

Reply:
Yes, you're right.